### PR TITLE
[5.9] Fixes isDirty() for MySQL JSON cast columns

### PR DESCRIPTION
**Pull request by nullthoughts**
As per @taylorotwell, resubmitting laravel/framework#29208 to Master branch with correct spacing in `originalIsEquivalent()`

---
Currently, `isDirty()` would return true for JSON cast columns with equal key-value pairs that are input in an order other than their random stored order. The reason for this false positive is because of MySQL's behavior with JSON column types:

[From the MySQL docs](https://dev.mysql.com/doc/refman/5.7/en/json.html#json-normalization):
> To make lookups more efficient, it also sorts the keys of a JSON object. You should be aware that the result of this ordering is subject to change and not guaranteed to be consistent across releases.

In plain English, this means that the array keys will often not be stored in the same order that they are input.

Behavior without the fix:
```
$model->jsonAttribute = [
    'foo' => 1,
    'bar' => 2,
    'baz' => 3,
];

$model->syncOriginal();

$model->jsonAttribute = [
    'baz' => 3,
    'foo' => 1,
    'bar' => 2,
];

$model->isDirty('jsonAttribute'); // true
```

The fix corrects this by checking for an empty result of `array_diff_assoc()` on JSON cast columns in `originalIsEquivalent()`. It also will fix the behavior for `getChanges()` (and naturally all other methods that depend on `getDirty()`).

### CODE REVIEW

**Comment by staudenmeir** [comment]
This doesn't work if you remove keys from the array:

```php
$user = User::create([
    'options' => ['a' => 'b', 'c' => 'd'],
]);

$user->update([
    'options' => ['a' => 'b'],
]);

dd($user->refresh()->options); // ['a' => 'b', 'c' => 'd']
```

**Comment by distinctm** [comment]
@staudenmeir Good catch, closing for now to rework

**Comment by staudenmeir** [comment]
It also doesn't work with nested arrays:

```php
$user = Use::create([
    'options' => ['a' => ['b' => 'c']],
]);

$user->update([
    'options' => ['a' => ['b' => 'c', 'd' => 'e']],
]);
```

This throws an exception:

> Array to string conversion

**Comment by distinctm** [comment]
@staudenmeir Correct, see above comment

**Comment by distinctm** [comment]
@staudenmeir I've made some changes by adding an `Arr::diff()` helper method. It instead of recursively comparing associative/multi-dimensional arrays, it is using the dot helper to flatten them before running `array_diff_assoc()`.

https://github.com/distinctm/framework/commit/78e34f874837b8a8a5090485790152943a58453e

I do feel that it might add too much overhead, but am open to suggestions. I'm also not a big fan of the `Arr::diff()` helper returning differences in "dot" notation. 

Additional tests have been implemented by adding a JSON specific model stub.

I'd love to hear your feedback on this approach. Again, it'd be nice to have support for `isDirty()` on JSON columns.

**Comment by staudenmeir** [comment]
The dot notation is ambiguous:

```php
$a = ['a.b' => 'c'];
$b = ['a' => ['b' => 'c']];
dd(Arr::diff($a, $b)); // []
```

**Comment by distinctm** [comment]
Meaning if you had array keys utilizing dots? `['a.b', 'h.k', 'm.n', ...]`

Then yes, it would present an issue. Although I do feel it's a bit of an edge case, I also am not 100% behind using flattened dot-arrays for comparison (although it is more elegant than a loop).

I'm open to any suggestions, I'll keep working on this as time permits. Feel free to reach out to me directly if you'd like to work on it together.

**Comment by staudenmeir** [comment]
> I do feel it's a bit of an edge case

It may be, but that doesn't matter here. We can't introduce new code that no longer detects certain changes.

I don't think there's a simple solution for this issue. Any approach would have to traverse both arrays.

**Comment by distinctm** [comment]
@staudenmeir I agree on both counts.

The following passes the Arr:diff tests (so far), but isn't very pretty or easy to understand.

```php
    /**
     * Returns bi-directional differences of two indexed, associative or multi-dimensional arrays.
     *
     * @param  array  $array1
     * @param  array  $array2
     * @return array
     */
    public static function diff(array $array1, array $array2)
    {
        if(!static::isAssoc($array1) && !static::isAssoc($array2)) {
            return array_merge(
                array_diff($array1, $array2),
                array_diff($array2, $array1)
            );
        }
        
        $keys = array_keys($array1);

        $items = array_map(function($value, $key) use ($array2) {
            if (is_array($value)) {
                if (!array_key_exists($key, $array2) || !is_array($array2[$key])) {
                    return $value;
                }
                
                return static::diff($value, $array2[$key]);
            }

            if (!array_key_exists($key, $array2) || $array2[$key] !== $value) {
                return $value;
            }
        }, $array1, $keys);
        
        return array_filter(
            array_combine($keys, $items)
        );
    }
```

**Comment by distinctm** [comment]
A foreach loop instead of the array_map could improve readability slightly

```php
    /**
     * Returns bi-directional differences of two indexed, associative or multi-dimensional arrays.
     *
     * @param  array  $array1
     * @param  array  $array2
     * @return array
     */
    public static function diff(array $array1, array $array2)
    {
        if(!static::isAssoc($array1) && !static::isAssoc($array2)) {
            return array_merge(
                array_diff($array1, $array2),
                array_diff($array2, $array1)
            );
        }
        
        $results = [];

        foreach ($array1 as $key => $value) {
            if (is_array($value)) {
                if (!array_key_exists($key, $array2) || !is_array($array2[$key])) {
                    $results[$key] = $value;
                } else {
                    $results[$key] = static::diff($value, $array2[$key]);
                }
            }
            
            if (!array_key_exists($key, $array2) || $array2[$key] !== $value) {
                $results[$key] = $value;
            }
        }

        return array_filter($results);
    }
```

**Comment by staudenmeir** [comment]
Cases that don't work:

```php
Arr::diff([[]], [[]]); // exception
Arr::diff(['a' => []], ['b' => []]); // []
```

**Comment by distinctm** [comment]
@staudenmeir Thanks for catching those, adding them to the tests. Looks like I'll need to remove array_filtering for the second one.

**Comment by Tofandel** [comment]
Wouldn't the [recursive iterator iterator](https://www.php.net/manual/en/class.recursiveiteratoriterator.php) be of help here?
Then you could just manually compare the keys in the traversal and exit early if something doesn't match you don't need to check if it's associative or sequential as they would have the same key values pairs in all cases
It would be performant and prevent other mentionned issues

**Comment by nullthoughts** [comment]
@Tofandel I haven't utilized RecursiveIteratorIterator previously, but will explore it some as time allows. It would be nice to resolve this (though I have to admit I have not dug deep to see if any under-the-hood changes in the new releases of Laravel help address it). Thanks for sharing your thoughts.
